### DISCUSSION TITLE
Teckin SP22 Smart Plug 10A 2200W with Power Measure

### DISCUSSION THREAD

**Discussion by xrust83** [RavenSystem/esp-homekit-devices#1814] [MEPLHAA Scripts Collection]
**Device Name:** Teckin SP22 Smart Plug 10A 2200W
**URL to Device:** https://www.evroz.com.ua/smart-rozetka-smart-plug-teckin-sp22-wifi.html
**Used HAA version:** 10.4.6
**Summary of functions:** Smart plug with LED, Button and Power Measure
**MEPLHAA:** `{"c":{"l":2,"n":"Smart Plug","b":[{"g":13,"t":5}]},"a":[{"0":{"r":[{"g":15},{"g":0,"v":1}]},"1":{"r":[{"g":15,"v":1},{"g":0}]},"t":2,"b":[{"g":13}],"es":[{"t":75,"j":10,"dt":[5,14,12],"vf":0.1536,"cf":0.0106,"pf":1.5957,"y0":[{"0":{"m":[[1]]},"v":245}],"y1":[{"0":{"m":[[1]]},"v":15}],"y2":[{"0":{"m":[[1]]},"v":2200}]}]}]}`

**Functions:** Switch on/off on single press. 
Power Measure in HAA Manager
Auto of else: U > 245V, C > 10A, P > 2200W
![IMG_4944](https://user-images.githubusercontent.com/58448955/163034209-7ce1f40c-01fe-496d-a1ca-d3079431980b.PNG)

**GPIO # | Component**

GPIO00 | LED red
GPIO02 | LED blue
GPIO05 | CF
GPIO10 | None
GPIO12 | SEL
GPIO13 | Button
GPIO14 | CF1
GPIO15 | Relay

![38102720-61ef963a-3384-11e8-92eb-da52fadc8db5](https://user-images.githubusercontent.com/58448955/163035351-2a387f83-033a-4d0d-a97b-369ceed2fb10.jpg)

**YouTube Video**
https://www.youtube.com/watch?v=2aregv5_mpU

**Comment by xrust83** [comment]
**Migration to HAA 12 MEPLHAA:** `{"c":{"io":[[[2],3],[[13],6,1],[[0,12,15],2],[[5,14],1]],"l":2,"n":"Smart-Plug","b":[{"g":13,"t":5}]},"a":[{"0":{"r":[{"g":15},{"g":0,"v":1}]},"1":{"r":[{"g":15,"v":1},{"g":0}]},"t":2,"b":[{"g":13}],"es":[{"t":75,"j":10,"dt":[5,14,12],"vf":0.1536,"cf":0.0106,"pf":1.5957,"y0":[{"0":{"m":[[1]]},"v":245}],"y1":[{"0":{"m":[[1]]},"v":15}],"y2":[{"0":{"m":[[1]]},"v":2200}]}]}]}`

**Reply by hkuehl** [reply]
hi @xrust83 I also tried your MEPLHAA for my SP22 running with 12.0.1.
I was able to add it to Homekit but the Switch itself is not performing any actions anymore. LED also stays dark. Pushing the physical button also does not perform any action. So I guess the MEPLHAA is wrong.
I guess it is an older model of the SP22. Do you have any insights in GPIO-mappings for those?

**Reply by xrust83** [reply]
> hi @xrust83 I also tried your MEPLHAA for my SP22 running with 12.0.1.
> 
> I was able to add it to Homekit but the Switch itself is not performing any actions anymore. LED also stays dark. Pushing the physical button also does not perform any action. So I guess the MEPLHAA is wrong.
> 
> I guess it is an older model of the SP22. Do you have any insights in GPIO-mappings for those?

this script worked with my outlet the first time. if teckin made changes to the board layout you can disassemble the socket and identify the gpio pins visually.

and what script did you take from the first post? or the new one posted below 6 days ago. in НAA 12 the appearance of the MEPLHAA script has changed

**Reply by hkuehl** [reply]
> and what script did you take from the first post? or the new one posted below 6 days ago. in НAA 12 the appearance of the MEPLHAA script has changed

I took the updated one you posted last.

I found [this](https://templates.blakadder.com/teckin_SP22.html)  mapping for an older SP22 from Tasmota-docs and you can see differences in GPIO mapping compared to your mapping. So I will update the MEPLHAA accordingly and try again.

**Reply by hkuehl** [reply]
I pretty-printed my config below with some comments. That one works for my older model. Do you have any idea what those wild-card-actions do?

And I am also only able to see the power measurements in Eve Home, not via Homekit.

```
{"c":{
    "io":[
        [[3],3],
        [[1],6,1],
        [[13,12,14],2],
        [[4,5],1]
    ],
    "l":3, #led on gpio 2
    "n":"Smart-Plug",
    "b":[{"g":1,"t":5}]}, # button on GPIO 13
    "a":[{
        "t":2, #service type Outlet
        "0":{"r":[{"g":14},{"g":13,"v":1}]}, #action definition for led when button pressed
        "1":{"r":[{"g":14,"v":1},{"g":13}]},
        "b":[{"g":1}], #binary input button
        "es":[{ #extra services
            "t":75, # type power measure
            "j":10, # period in seconds between readings
            "dt":[4,5,12], #chip data: GPIO [CF, CF1, Sel]
            "vf":0.1536,"cf":0.0106,"pf":1.5957,
            "y0":[{"0":{"m":[[1]]},"v":245}], # what do these actions do?
            "y1":[{"0":{"m":[[1]]},"v":15}], # what do these actions do?
            "y2":[{"0":{"m":[[1]]},"v":2200}] # what do these actions do?
            }]
        }]}
```

**Reply by RavenSystem** [reply]
Those wildcard actions will turn off outlet if one of these cases is done:
- 245V are reached.
- 15A are reached.
- 2200W are reached.

**Reply by xrust83** [reply]
> I pretty-printed my config below with some comments. That one works for my older model. Do you have any idea what those wild-card-actions do?
> 
> And I am also only able to see the power measurements in Eve Home, not via Homekit.
> 
> ```
> {"c":{
>     "io":[
>         [[3],3],
>         [[1],6,1],
>         [[13,12,14],2],
>         [[4,5],1]
>     ],
>     "l":3, #led on gpio 2
>     "n":"Smart-Plug",
>     "b":[{"g":1,"t":5}]}, # button on GPIO 13
>     "a":[{
>         "t":2, #service type Outlet
>         "0":{"r":[{"g":14},{"g":13,"v":1}]}, #action definition for led when button pressed
>         "1":{"r":[{"g":14,"v":1},{"g":13}]},
>         "b":[{"g":1}], #binary input button
>         "es":[{ #extra services
>             "t":75, # type power measure
>             "j":10, # period in seconds between readings
>             "dt":[4,5,12], #chip data: GPIO [CF, CF1, Sel]
>             "vf":0.1536,"cf":0.0106,"pf":1.5957,
>             "y0":[{"0":{"m":[[1]]},"v":245}], # what do these actions do?
>             "y1":[{"0":{"m":[[1]]},"v":15}], # what do these actions do?
>             "y2":[{"0":{"m":[[1]]},"v":2200}] # what do these actions do?
>             }]
>         }]}
> ```

"vf":0.1536,"cf":0.0106,"pf":1.5957,  

this is the calibration of my outlet, you need to calibrate on your particular outlet
